### ISSUE TITLE
[AI] flaggems route: flag_gems index_select returns partially-unwritten output in HF cached beam search on DCU

### ISSUE THREAD

**Issue by ethanZZZZZZZZZZZZ** [flagos-ai/Torch-FL#120]
---
name: AI Agent Issue
about: Issue created by Claude Code or other AI coding agents
title: "[AI] flaggems route: flag_gems index_select returns partially-unwritten output in HF cached beam search on DCU"
labels: ai-generated
assignees: ''
---

## Issue Type
- [x] Bug Report

## AI Agent Information
- **Agent**: Claude Code CLI
- **Model**: Claude (Kimi-K3 harness)
- **Session Context**: HF v5.5.0 official UT sweep of 5 embedding-model families
  on the DCU flaggems route; 3 beam-search UT failures across T5/Qwen3/Gemma3
  share this root cause. Full report: `docs/bench_embedding_families_ut_three_routes.md`.

## Summary

On the DCU flaggems route, `aten::index_select` (routed to `flagos_python`)
intermittently returns an output whose tail was **never written** (stale
`torch.empty` bytes). In HF cached beam search
(`DynamicCache.reorder_cache` → `index_select(0, beam_idx)`) the kernel reads
the freshly-produced `beam_idx` before it lands on device, judges one index
out-of-range, and skips that output column — poisoning the KV cache with
garbage/NaN. The flag_gems kernel itself is numerically correct (standalone
stress clean); the defect is a missing stream-ordering edge between
flag_gems' triton launch and the flagos (PrivateUse1) current stream.

## Environment
<details>
<summary>Click to expand</summary>

- **Platform**: DCU (hygon), `HIP_VISIBLE_DEVICES=0`
- **Python/PyTorch**: 3.10 / 2.10.0+cpu frontend + bundled DTK libtorch
- **torch_fl**: 0.1.0+dtk, origin/main @ `39b15c0`
- **flag_gems**: 5.4.0.dev0 (hygon) + DTK triton (`TRITON_BACKENDS_IN_TREE=1`)
- **Runtime**: `TRANSFORMERS_TEST_DEVICE=flagos FLAGOS_USE_FLAGGEMS=1`,
  conf `torch_fl/configs/backends_dcu_flaggems.conf` (`index_select = flagos_python`)
</details>

## Reproduction
```python
import torch, torch_fl
from transformers import T5Config, T5ForConditionalGeneration

config = T5Config(vocab_size=99, d_model=32, d_kv=4, d_ff=37, num_layers=2,
                  num_decoder_layers=2, num_heads=4, dropout_rate=0.1,
                  layer_norm_epsilon=1e-6, feed_forward_proj="relu",
                  is_encoder_decoder=True, pad_token_id=0, eos_token_id=1,
                  decoder_start_token_id=0)
torch.manual_seed(0)
model = T5ForConditionalGeneration(config=config).to("flagos").eval()
torch.manual_seed(42)
ids = torch.randint(0, 99, (13, 7), device="flagos")
print(model.generate(ids[:1], num_beams=2, max_length=5,
                     do_sample=False, use_cache=True).tolist())
```

Failing HF UT nodes (all pass on vendor `cuda` baseline):
`T5ModelTest::test_generate_with_past_key_values` (deterministic),
`Qwen3ModelTest::test_generate_from_inputs_embeds_1_beam_search` (flaky),
`Gemma3Vision2TextModelTest::test_generate_from_inputs_embeds_1_beam_search`
(deterministic).

## Expected vs Actual Behavior
**Expected:** `[[0, 33, 33, 33, 33]]` (matches boxing route / vendor / `use_cache=False`)
**Actual:** `[[0, 0, 32]]` — NaN logits from decoder step 1.

Byte-level capture of the first corrupt call (input `(2,4,1,4)` cache keys,
index `[0,0]`): output's second index column = `[0.0×10, 6 garbage floats]`
(stale recycled-block bytes); an identical relaunch after a D2H drain is
fully correct.

## Root Cause Analysis

Call chain: `BeamSearchScorer` (produces `beam_idx` on flagos stream) →
`DynamicCache.reorder_cache` (`transformers/cache_utils.py:81-85`) →
`aten::index_select` → `flag_gems.ops.index_select`
(`ops/index_select.py:54` `torch.empty` + `:64` kernel launch).

The triton kernel (`ops/index_select.py:17-38`) reads the device `beam_idx`
with no ordering against the flagos stream that just produced it; a stale
index entry fails the `indices < N` validity mask, so that output column is
never stored and keeps recycled `torch.empty` bytes.

Evidence: (1) loss is exactly one index column across all grid tiles →
mask-granular, not tile-granular; (2) host-side index snapshots stable
(`[0,0]`); (3) standalone stress (same shapes, contig/non-contig, duplicate
indices) 0 mismatches → kernel math ruled out; (4) per-op
`torch.cuda.synchronize()` does not heal it (does not cover flagos queues),
while a `.cpu()` round-trip does → ordering, not arithmetic; (5) op-level CPU
shadow flags `index_select` as the first/only diverging op; (6) routing only
`index_select = cuda` cures the reproducer 3/3 and all three UT nodes.

Classification: cross-stream data race (missing happens-before) in the
torch_fl flagos ↔ flag_gems triton launch integration — **not** a flag_gems
arithmetic bug (unlike the previously root-caused `layer_norm_backward` M/N
kernel defect, and distinct from #86/#100 `index.Tensor`).

## Proposed Solution

- **Mitigation (verified)**: route `index_select`/`index_select.out` to `cuda`
  in `backends_dcu_flaggems.conf` until fixed.
- **Real fix (torch_fl runtime)**: order flag_gems kernel launches with the
  flagos current stream — record/wait an event around each triton launch, or
  make the hcu backend's `get_current_stream()` return the flagos current
  stream for PrivateUse1 devices; also cover flagos queues in
  `torch.cuda.synchronize()`.
- **Audit**: same race window applies to other flag_gems kernels taking
  freshly-produced small device args (`gather`, `topk`, `cumsum`, …).

## Verification Plan
- [ ] Unit: T5 tiny `generate(num_beams=2, use_cache=True)` == `use_cache=False` output.
- [ ] Unit: 1000× in-flight `index_select` on 4-D cache shapes vs CPU reference.
- [ ] Integration: the three failing HF UT nodes pass on the flaggems route.
- [ ] Manual: full T5/Qwen3/Gemma3 suites back to non-G1 baseline.

## Context & Investigation

Artifacts under `/root/bench/`: NaN reproducer + byte capture
(`repro_t5_nan.py`, `repro_t5_iswrap.py`, `repro_t5_isidx.py`,
`repro_t5_isdump.py`), op-level CPU shadow (`repro_t5_opshadow.py`),
standalone stress (`repro_is4d*.py`), RNG/zeros exclusions
(`repro_t5_rng.py`, `repro_t5_zeros.py`), conf bisect
(`bisect_f2/combo_no_index_select.conf`), UT logs
(`results/{t5,qwen3,gemma3}_flaggems.log`).

## Related Code Locations
- `flag_gems/ops/index_select.py:17-38, 41-71` — kernel + wrapper
- `flag_gems/runtime/backend/_hygon/heuristics_config_utils.py:79-85` — grid heuristics
- `transformers/cache_utils.py:81-85` — `DynamicCache.reorder_cache`
- `torch_fl/configs/backends_dcu_flaggems.conf` — routing entry (mitigation)
- `triton/backends/hcu/driver.py` — launch/stream acquisition (fix candidate)

## Checklist - AI Agents MUST Complete All
- [x] Complete environment information
- [x] Minimal, self-contained reproducer
- [x] Full error output (byte-level capture)
- [x] Root cause analyzed (not just symptoms)
- [x] Specific solution proposed
- [x] Affected code locations with line numbers
- [x] Fix verification described
- [x] Checked for duplicate issues (#86/#100, layer_norm M/N — different defects)
- [x] All text in **English**
